#748 Match Route configurations for upstream paths when empty Catch-All placeholders at the end of template

AlyHKafoury opened the pull request:
## Fixes #748 
  - #748

raman-m commented on test/Ocelot.AcceptanceTests/RoutingTests.cs:
```diff
@@ -1,5 +1,5 @@
-using Microsoft.AspNetCore.Http;
-using Ocelot.Configuration.File;
+using Microsoft.AspNetCore.Http;
+using Ocelot.Configuration.File;
```
Zero diff!
Line endings?
What IDE do you use to write C# code?
Please try to remove wrong line endings. So, don't change them!

AlyHKafoury replied:
Using Vscode

raman-m commented on test/Ocelot.AcceptanceTests/RoutingTests.cs:
```diff
@@ -483,46 +483,46 @@ public void should_return_ok_when_upstream_url_ends_with_forward_slash_but_templ
-                .And(x => _steps.ThenTheResponseBodyShouldBe("Hello from Laura"))
-                .BDDfy();
+
+        [Theory]
+        [InlineData("/account/authenticate")]
+        [InlineData("/account/authenticate/")]
+        [Trait("Issue", "649")]
+        public void should_fix_issue_649(string url)
+        {
+            var port = PortFinder.GetRandomPort();
+            var baseUrl = $"http://localhost:{port}";
+            var configuration = new FileConfiguration
```
Seems entire test body has no changes , but all these lines have been included into diff.
Please, care about old line endings or indents!

AlyHKafoury commented:
@raman-m I believe the PR is ready for review please give me your thoughts about this, and if there is anything I need to improve

AlyHKafoury commented on test/Ocelot.UnitTests/Configuration/UpstreamTemplatePatternCreatorTests.cs:
@raman-m the changes in here reflect the Regex changes from matching 1 or more characters to zero or more characters

raman-m replied:
All these tests below have failed when empty placeholder? Or they failed because of changed RegEx which allows empty placeholder?

Well... we have to double check everything... if we've changed internal RegEx in service class.
Seems this new updated RegEx touches a lot of logic when placeholder is empty.
That will be massive logic upgrade and we need to inform developers that empty placeholders are allowed everywhere now.

AlyHKafoury replied:
Both reasons

AlyHKafoury commented on test/Ocelot.AcceptanceTests/RoutingTests.cs:
```diff
@@ -589,7 +589,8 @@ public void should_return_not_found()
                 .When(x => _steps.WhenIGetUrlOnTheApiGateway("/products/"))
-                .Then(x => _steps.ThenTheStatusCodeShouldBe(HttpStatusCode.NotFound))
+                .Then(x => _steps.ThenTheStatusCodeShouldBe(HttpStatusCode.OK))
+                .Then(x => ThenTheDownstreamUrlPathShouldBe("/products"))
```
@raman-m the change in this test reflects the correct desired behavior of matching empty places holders if they appear  on upstream path and there is a route that matches

raman-m replied:
You need to convert this `Fact` to `Theory`

raman-m commented on src/Ocelot/DownstreamRouteFinder/UrlMatcher/UrlPathPlaceholderNameAndValueFinder.cs:
```diff
@@ -50,11 +50,16 @@ public Response<List<PlaceholderNameAndValue>> Find(string path, string query, s
                     return new OkResponse<List<PlaceholderNameAndValue>>(placeHolderNameAndValues);
                 }
-                else if (IsCatchAll(path, counterForPath, pathTemplate))
+                else if (IsCatchAll(path, counterForPath, pathTemplate) || (pathTemplate[counterForTemplate] == '{') && NoMoreForwardSlash(pathTemplate, counterForTemplate) && NotPassedQueryString(pathTemplate, pathTemplate.Length))
```
Why did you add this long and complex condition here just for user cases?

Which paths will be handled here?

AlyHKafoury replied:
This Logic here is the one that will handle the case of a catch all in the end of line and handles it with the same logic as the normal catchall

raman-m commented on src/Ocelot/DownstreamRouteFinder/UrlMatcher/UrlPathPlaceholderNameAndValueFinder.cs:
```diff
@@ -50,11 +50,16 @@ public Response<List<PlaceholderNameAndValue>> Find(string path, string query, s
                     var endOfPlaceholder = GetNextCounterPosition(pathTemplate, counterForTemplate, '}');
 
-                    var placeholderName = GetPlaceholderName(pathTemplate, 1);
+                    if (pathTemplate[counterForTemplate] == '/')
+                    {
+                        counterForTemplate++;
+                    }
```
What do you fix here?
What case with what path?

AlyHKafoury replied:
this is the Case when to handle the following 
/{placeholder} => /{placeholder} 
to handle the case of : /{emptystring} => /{emptystring}

raman-m commented on src/Ocelot/DownstreamRouteFinder/UrlMatcher/UrlPathPlaceholderNameAndValueFinder.cs:
```diff
@@ -104,6 +109,11 @@ private static bool NothingAfterFirstForwardSlash(string path)
 
         private static string GetPlaceholderValue(string urlPathTemplate, string query, string variableName, string urlPath, int counterForUrl, char delimiter)
         {
+            if(counterForUrl > urlPath.Length)
+            {
+                return "";
+            }
```
I that a border case?
Which one?
If counter reached the end of url then no plaveholder values?

AlyHKafoury replied:
Yes exactly this is the way to circuit break any empty value in the end of the path for all the cases

raman-m commented on test/Ocelot.AcceptanceTests/RoutingTests.cs:
```diff
@@ -806,7 +807,42 @@ public void should_return_response_200_with_placeholder_for_final_url_path()
+            this.Given(x => GivenThereIsAServiceRunningOn($"http://localhost:{port}", "/", 200, "Test Body"))
+                .And(x => _steps.GivenThereIsAConfiguration(configuration))
+                .And(x => _steps.GivenOcelotIsRunning())
+                .When(x => _steps.WhenIGetUrlOnTheApiGateway("/upstream/test/"))
+                .Then(x => ThenTheDownstreamUrlPathShouldBe("/downstream/test/"))
```
Please include the case for `/upstream/test` (without ending slash)

raman-m commented on src/Ocelot/Configuration/Creator/UpstreamTemplatePatternCreator.cs:
```diff
@@ -5,7 +5,7 @@ namespace Ocelot.Configuration.Creator
     public class UpstreamTemplatePatternCreator : IUpstreamTemplatePatternCreator
     {
-        private const string RegExMatchOneOrMoreOfEverything = ".+";
+        private const string RegExMatchZeroOrMoreOfEverything = ".*";
```
Seems we need to make this constant as `public` because we need it for tests.
Otherwise you have to care about them manually.
We have to add /embed this regex value as a part of testing templates to show a reader (code reviewer) that those test are dependent much from this RegEx. It will be easy to read and support the tests in future.

AlyHKafoury replied:
@raman-m you are correct I will make it public because all the test cases use this

raman-m reviewed (changes requested):
Let's make it better.

AlyHKafoury commented:
@raman-m I addressed most of your concerns, and Looking forward to your new comments

raman-m commented:
Thanks for working on code review issues!
Too busy!... Sorry... Going to ask another Ocelot team members to review..

**FYI** I've added this PR to **Jan'24** milestone...
If you are on fire having desire to work more, please pick up open tickets from [Annual'23](https://github.com/ThreeMammals/Ocelot/milestone/3) backlog!... But seems there are no free tickets to assign because all Annual tickets have their PRs... Well... We can find something else...

raman-m commented:
@ggnaegi @RaynaldM  Please review!

AlyHKafoury commented:
> Thanks for working on code review issues!
> Too busy!... Sorry... Going to ask another Ocelot team members to review..
> 
> **FYI** I've added this PR to **Jan'24** milestone...
> If you are on fire having desire to work more, please pick up open tickets from [Annual'23](https://github.com/ThreeMammals/Ocelot/milestone/3) backlog!... But seems there are no free tickets to assign because all Annual tickets have their PRs... Well... We can find something else...

Thanks alot, I am gonna lookup for More tickets to work on I am really excited to contribute more

raman-m commented on test/Ocelot.UnitTests/Configuration/UpstreamTemplatePatternCreatorTests.cs:
```diff
@@ -42,7 +42,7 @@ public void should_use_re_route_priority()
             this.Given(x => x.GivenTheFollowingFileRoute(fileRoute))
                 .When(x => x.WhenICreateTheTemplatePattern())
-                .Then(x => x.ThenTheFollowingIsReturned("^(?i)/orders/.+$"))
+                .Then(x => x.ThenTheFollowingIsReturned($"^(?i)/orders(?:|/{UpstreamTemplatePatternCreator.RegExMatchZeroOrMoreOfEverything})$"))
```
Quite long sentence which hard to read! 
```suggestion
                private const string MatchEverything = UpstreamTemplatePatternCreator.RegExMatchZeroOrMoreOfEverything;
                .Then(x => x.ThenTheFollowingIsReturned($"^(?i)/orders(?:|/{MatchEverything})$"))
```

raman-m commented on test/Ocelot.UnitTests/Configuration/UpstreamTemplatePatternCreatorTests.cs:
```diff
@@ -229,7 +229,7 @@ public void should_create_template_pattern_that_matches_query_string_with_multip
             this.Given(x => x.GivenTheFollowingFileRoute(fileRoute))
                 .When(x => x.WhenICreateTheTemplatePattern())
-                .Then(x => x.ThenTheFollowingIsReturned("^(?i)/api/subscriptions/[^/]+/updates\\?unitId=.+&productId=.+$"))
+                .Then(x => x.ThenTheFollowingIsReturned("^(?i)/api/subscriptions/[^/]+/updates\\?unitId=.*&productId=.*$"))
```
Construct expected reg-expression with `MatchEverything` constant.

raman-m reviewed (changes requested):
@AlyHKafoury commented [on Jan 11](https://github.com/ThreeMammals/Ocelot/pull/1911#issuecomment-1885927140)
> I addressed most of your concerns, and Looking forward to your new comments

Wow! Quite quick and good fixes!

----
Here code review 2nd round! 😄 
Clean code, styling and code smells.

AlyHKafoury commented on src/Ocelot/DownstreamRouteFinder/UrlMatcher/UrlPathPlaceholderNameAndValueFinder.cs:
```diff
@@ -97,13 +97,27 @@ private static bool IsCatchAll(string path, int counterForPath, string pathTempl
                      && pathTemplate.IndexOf('}') == pathTemplate.Length - 1;
         }
 
+        private static bool IsCatchAllAfterOtherPlaceholders(string pathTemplate, int counterForTemplate)
```
will add a unit test for it

AlyHKafoury commented on src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs:
```diff
@@ -32,7 +32,8 @@ public class DownstreamUrlCreatorMiddleware : OcelotMiddleware
         {
-            var downstreamRoute = httpContext.Items.DownstreamRoute();
+            var downstreamRoute = httpContext.Items.DownstreamRoute();
+            var upstreamPath = httpContext.Request.Path.ToString();
```
I need access to the path to determine if I should return the slash or not based on is it included in the __Request Path__ or not

AlyHKafoury commented:
@raman-m my daily submission, I addressed most of the comments regarding styling and clean code. waiting for your review

raman-m commented:
@AlyHKafoury  Did you check all user scenarios from #748 by the tests?

AlyHKafoury commented:
> @AlyHKafoury Did you check all user scenarios from #748 by the tests?

Yes these are all the requests on the issue and extra ones for other cases I found that needed to be verified

raman-m commented:
OK I'm going to check code coverage and do quick final code review...

AlyHKafoury commented:
@raman-m if you have any changes you can request I am looking forward to contribute more in a lot of tickets

raman-m reviewed (changes requested):
Sequence Coverage: [0.8481](https://app.circleci.com/pipelines/github/ThreeMammals/Ocelot/1634/workflows/bd2fed33-ecc5-427e-b976-5346f9cf64df/jobs/3112?invite=true#step-102-816)
Total Line coverage: 85.9%
Total Branch coverage: 77.4%
- 100% + 100% for `Ocelot.Configuration.Creator.UpstreamTemplatePatternCreator` including lines 58-63 ✔️ 
- 96.9% + 96% for `Ocelot.DownstreamRouteFinder.UrlMatcher.UrlPathPlaceholderNameAndValueFinder` where lines 22, 51, **114-118 are not covered**❗ But lines 100-105 are covered ✔️ 
- 93% + 82.1% for `Ocelot.DownstreamUrlCreator.Middleware.DownstreamUrlCreatorMiddleware` where lines **42-53 are not covered**❗including **new lines 50-53 are not covered**❗ 

```diff
! Please write more unit tests for newly added lines!
```

raman-m commented on test/Ocelot.UnitTests/DownstreamRouteFinder/UrlMatcher/UrlPathPlaceholderNameAndValueFinderTests.cs:
```diff
@@ -349,6 +349,20 @@ public void can_match_down_stream_url_with_downstream_template_with_place_holder
                  .BDDfy();
         }
 
+        [Fact]
+        public void check_for_placeholder_at_end_of_template() 
+        {
+            var expectedTemplates = new List<PlaceholderNameAndValue>
+            {
+                new("{testId}", string.Empty),
+            };
+            this.Given(x => x.GivenIHaveAUpstreamPath("/upstream/test/"))
+                .And(x => x.GivenIHaveAnUpstreamUrlTemplate("/upstream/test/{testId}"))
+                .When(x => x.WhenIFindTheUrlVariableNamesAndValues())
+                .And(x => x.ThenTheTemplatesVariablesAre(expectedTemplates))
+                .BDDfy();
+        }
```
I know other unit tests can cover our user scenarios, and acceptance tests can cover too, **but...**
For bugs it is better to write one `Theory` test with name `should_fix_748_issue` with all templates from descriptions, author's and contributor's ones.

### User scenario 1

```
"DownstreamPathTemplate": "/api/invoices/{url}"
"UpstreamPathTemplate": "/{url}"
```

It maps URLs this way:

```
/ → /api/invoices/
/123 → /api/invoices/123
```

### User scenario 2

```
"DownstreamPathTemplate": "/api/invoices/{url}"
"UpstreamPathTemplate": "/test/{url}"
```

It works correctly when `{url}` is specified: `/test/123 → /api/invoices/123`

But it fails, when `{url}` is empty. I would expect upstream URL `/test/` to reroute to downstream URL `/api/invoices/`.

### User scenario 3

```
"DownstreamPathTemplate": "/api/invoices{url}"
"UpstreamPathTemplate": "/test{url}"
```

It correctly maps non epmty `{url}` which means, it works with the slash at the end, because the slash is the value of the `{url}`. But still not without it:

```
/test/123 → /api/invoices/123
/test/ → /api/invoices/
/test → Failed to match ReRoute configuration for upstream path: /test
```

To make work everything what I want, I need three reroutes (or at least two, if the first one will be without slashes):

```
{
    "DownstreamPathTemplate": "/api/invoices/{url}"
    "UpstreamPathTemplate": "/test/{url}"
},
{
    "DownstreamPathTemplate": "/api/invoices/"
    "UpstreamPathTemplate": "/test/"
},
{
    "DownstreamPathTemplate": "/api/invoices"
    "UpstreamPathTemplate": "/test"
}
```
See [original description](https://github.com/ThreeMammals/Ocelot/issues/748#issue-399043741).

raman-m replied:
For placeholders in the middle, please write one extra `Theory` with unit tests 👉 
> [satano](https://github.com/satano) commented [on Jan 15, 2019](https://github.com/ThreeMammals/Ocelot/issues/748#issuecomment-454305176)

AlyHKafoury replied:
@raman-m all the unit tests exist as part of the routing tests
https://github.com/ThreeMammals/Ocelot/pull/1911/files#diff-99c1f697736fe2348d794d9911adf9dde5428205b123b99511fab187260ffd60L810

raman-m reviewed (changes requested):
More unit tests required with test data from original descriptions
- [satano](https://github.com/satano) commented [on Jan 14, 2019](https://github.com/ThreeMammals/Ocelot/issues/748#issue-399043741)
- [satano](https://github.com/satano) commented [on Jan 15, 2019](https://github.com/ThreeMammals/Ocelot/issues/748#issuecomment-454305176)

Please, write two theories!

AlyHKafoury commented:
@raman-m is there a way I can run test coverage myself ?

raman-m commented:
Tomorrow...
Have a good evening!

AlyHKafoury commented:
@raman-m unit tests are added for the code, but I cannot check coverage yet

raman-m commented:
@AlyHKafoury commented [on Jan 15](https://github.com/ThreeMammals/Ocelot/pull/1911#issuecomment-1892634268):
> Is there a way I can run test coverage myself?

Sure, there is! We use [coverlet.collector](https://github.com/search?q=repo%3AThreeMammals%2FOcelot%20coverlet.collector&type=code) package (see [NuGet](https://www.nuget.org/packages/coverlet.collector)) to run code coverage by [Coverlet](https://github.com/coverlet-coverage/coverlet) tool. This is standard tool which included in xUnit Test Project template by Visual Studio. For NUnit template I'm not sure.

To get coverage results you need to
- open console, Cmd or Bash
- go to your repo and UnitTesting folder, for example in my environment: `D:\github\Ocelot\AlyHKafoury\Ocelot\test\Ocelot.UnitTests`
- run command: `dotnet test --collect:"XPlat Code Coverage"`
- wait for 1-2 minutes while building & testing, and you'll get a log like this:
  ![image](https://github.com/ThreeMammals/Ocelot/assets/10501504/ba925c63-d214-45c1-977f-da4179e24a21)
- copy the last GUID from the results list, this is GUID of .NET 8 case running.
- make sure that command `ReportGenerator` is working in console, otherwise install **dotnet-reportgenerator-globaltool** by command `dotnet tool install --global dotnet-reportgenerator-globaltool --version 5.2.0`. See [NuGet](https://www.nuget.org/packages/dotnet-reportgenerator-globaltool)
- run command with replacement of your copied GUID:
  `reportgenerator -reports:"d:\github\Ocelot\AlyHKafoury\Ocelot\test\Ocelot.UnitTests\TestResults\57b3a0c8-b218-4197-a5e2-b20fe6164078\coverage.cobertura.xml" -targetdir:"d:\github\Ocelot\AlyHKafoury\Ocelot\test\Ocelot.UnitTests\TestResults" -reporttypes:Html`.
  Sure you have to replace GUID and adjust your path to the solution folder.
- you'll get a log like this:
  ![image](https://github.com/ThreeMammals/Ocelot/assets/10501504/cc0d5599-9708-4400-aff3-5ac3c067dad5)
- go to `d:\github\Ocelot\AlyHKafoury\Ocelot\test\Ocelot.UnitTests\TestResults` folder and open `index.html` file in your browser

Sure your `TestResults` folder can be located anywhere.
Hope it helps!

raman-m commented on test/Ocelot.AcceptanceTests/RoutingTests.cs:
```diff
@@ -808,6 +810,43 @@ public void should_return_response_200_with_placeholder_for_final_url_path()
+        [InlineData("/downstream/test/{testId}", "/upstream/test/{testId}", "/upstream/test123", null)]
+        [InlineData("/downstream/{version}/test", "/upstream/{version}/test", "/upstream/v1/test", "/downstream/v1/test")]
+        [InlineData("/downstream/{version}/test", "/upstream/{version}/test", "/upstream/test", null)]
+        public void should_return_correct_downstream_when_omitting_ending_placeholder(string downstreamPathTemplate, string upstreamPathTemplate, string requestURL, string downstreamURL) //should_fix_748_issue
```
Very good!
Let me to do final review...

raman-m commented on src/Ocelot/DownstreamRouteFinder/UrlMatcher/UrlPathPlaceholderNameAndValueFinder.cs:
```diff
@@ -97,13 +97,25 @@ private static bool IsCatchAll(string path, int counterForPath, string pathTempl
+                && (counterForTemplate < pathTemplate.Length - 1)
+                && (pathTemplate[counterForTemplate + 1] == '{')
+                && NoMoreForwardSlash(pathTemplate, counterForTemplate + 1)
+                && NotPassedQueryString(pathTemplate, pathTemplate.Length);
```
2 tests for
- CatchAll at the end
- CatchAll at the end after a placeholder in the middle

AlyHKafoury replied:
@raman-m added extra test cases in here https://github.com/ThreeMammals/Ocelot/pull/1911/files#diff-fe239522803966c25dc7eda0de976c6b65aedb1d095c709cf3e414aa06f62f4cR389

AlyHKafoury replied:
> I'm sorry! What URL paths do we process here? Like this? 👇
> 
> ```
> "DownstreamPathTemplate": "/api/{version}/test/{url}"
> "UpstreamPathTemplate": "/test/{version}/{url}"
> ```
> 
> According to our [Placeholders](https://ocelot.readthedocs.io/en/latest/features/routing.html#placeholders:~:text=This%20will%20forward%20any%20path%20%2B%20query%20string%20combinations%20to%20the%20downstream%20service%20after%20the%20path%20%2Fapi.) docs and [Catch All](https://ocelot.readthedocs.io/en/latest/features/routing.html#id1) docs, we have to forward query string too ❗ **We must forward query string in case of empty Catch All placeholders too!** Seems we have to remove the line 105.
> 
> Could you add extra 1 unit and 1 acceptance tests which check the case with query string too, please?

we don't have to remove that line because this branch comes into effect when there is a specific condition of no query strings till the current PathCounter other than that it is handled normally but the if branch not the else if branch

raman-m commented on src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs:
```diff
@@ -68,9 +75,9 @@ public async Task Invoke(HttpContext httpContext)
-                    downstreamRequest.Query = string.IsNullOrEmpty(downstreamRequest.Query)
-                        ? newQuery
-                        : MergeQueryStringsWithoutDuplicateValues(downstreamRequest.Query, newQuery, placeholders);
+                    downstreamRequest.Query = string.IsNullOrEmpty(downstreamRequest.Query)
+                        ? newQuery
+                        : MergeQueryStringsWithoutDuplicateValues(downstreamRequest.Query, newQuery, placeholders);
```
Zero diff! No code changes. But GitHub highlighted!
Line endings?
**Please, use Visual Studio!** We have special editor settings for mixed line endings!
Please rebase onto develop, which should help to remove ghost changes.

AlyHKafoury replied:
I am using Visual studio, and line-ending in all files are warned they are mixed, I believe we need to make a pass on line endings. but I will rebase to fix this

raman-m replied:
Yes, rebase please! It should help to remove unclear changes.

raman-m commented on test/Ocelot.AcceptanceTests/RoutingTests.cs:
```diff
@@ -808,6 +810,44 @@ public void should_return_response_200_with_placeholder_for_final_url_path()
+        [InlineData("/downstream/test/{testId}", "/upstream/test/{testId}", "/upstream/test123", null)]
+        [InlineData("/downstream/{version}/test/{url}", "/upstream/{version}/test/{url}", "/upstream/v1/test/123", "/downstream/v1/test/123")]
+        [InlineData("/downstream/{version}/test", "/upstream/{version}/test", "/upstream/v1/test", "/downstream/v1/test")]
+        [InlineData("/downstream/{version}/test", "/upstream/{version}/test", "/upstream/test", null)]
```
😮 I wonder why `string downstreamURL` is `null` here? 😵 
Should not it be `"/downstream/test"` because of empty `{version}` placeholder?

AlyHKafoury replied:
is null because there is no match in here because the placeholder is in the middle, that is a requirement on the issue

raman-m commented on test/Ocelot.UnitTests/DownstreamRouteFinder/UrlMatcher/UrlPathPlaceholderNameAndValueFinderTests.cs:
```diff
@@ -349,6 +349,38 @@ public void can_match_down_stream_url_with_downstream_template_with_place_holder
+        }
+
+        [Theory]
+        [InlineData("/api/invoices/{url}", "/api/invoices/123", "{url}", "123")]
+        [InlineData("/api/invoices/{url}", "/api/invoices/", "{url}", "")]
+        [InlineData("/api/invoices/{url}", "/api/invoices", "{url}", "")]
+        [InlineData("/api/{version}/invoices/", "/api/v1/invoices/", "{version}", "v1")]
```
Query string is not checked!
But I'm not sure this is the right place to test query strings.
Let's try!...

raman-m commented on test/Ocelot.UnitTests/DownstreamUrlCreator/DownstreamUrlCreatorMiddlewareTests.cs:
```diff
@@ -501,6 +501,44 @@ public void Should_forward_query_parameters_without_duplicates(string everything
         }
 
+        [Theory]
+        [InlineData("/test/{version}/{url}", "/api/{version}/test/{url}", "/test/v1/123", "{url}", "123", "/api/v1/test/123")]
+        [InlineData("/test/{version}/{url}", "/api/{version}/test/{url}", "/test/v1/", "{url}", "", "/api/v1/test/")]
+        [InlineData("/test/{version}/{url}", "/api/{version}/test/{url}", "/test/v1", "{url}", "", "/api/v1/test")]
```
Seems here we have to check query strings too.

raman-m reviewed (changes requested):
We need to polish tests!

AlyHKafoury commented:
I've added test cases for query strings and in middle replacements @raman-m

AlyHKafoury commented:
@raman-m should I fix https://github.com/ThreeMammals/Ocelot/commit/9f7e640ae8efdb3ff1e2325ce5e3c52c8105fa82

raman-m commented:
I've debugged a little bit, seems the problem is [here](https://github.com/ThreeMammals/Ocelot/pull/1911/files#diff-611108f6b3268be0a8735448eb7b2b23e994a1bbf18b4a99794a9f720cdd121aR61-R63)

In feature branch: [lines 60-64](https://github.com/AlyHKafoury/Ocelot/blob/Fix-748/src/Ocelot/Configuration/Creator/UpstreamTemplatePatternCreator.cs#L60-L64)

AlyHKafoury commented:
@raman-m something is broken with Circle-CI ?

AlyHKafoury commented:
@raman-m all tests are passing locally now, please retry the Circle-CI test and check if you need more changes or unit tests

raman-m commented:
> @raman-m something is broken with Circle-CI ?

Why do you think so?

AlyHKafoury commented:
@raman-m the tests fail because being unable to pull branch

raman-m commented:
Wow! Maybe CircleCI netwotking issue or GitHub deployments...

raman-m commented:
Everything is fine with your GitHub repo... I've re-cloned successfully...
Let me run the CircleCI build once again...

raman-m commented:
🆗 Let's wait the build 10 minutes... 😃  It was environment issue on CircleCI side.
Are tests green locally?

AlyHKafoury commented:
@raman-m tests have passed on Circle-CI, please do inform me if there are more needed changes

raman-m commented:
Well, Aly... I've added https://github.com/ThreeMammals/Ocelot/pull/1911/commits/2f614565564b68b68557399208ef7bf64ef06e11 which will have failed build because of Routing Tests when `string downstreamURL` is `null` while checking by [these lines 851-856](https://github.com/ThreeMammals/Ocelot/pull/1911/files#diff-99c1f697736fe2348d794d9911adf9dde5428205b123b99511fab187260ffd60R851-R856).
So, I added them to check query string case.
Probably it is better to move query constants to `InlineData` as extra parameter.

Please, fix and this will be the end of development. 😋

AlyHKafoury commented:
@raman-m I fixed the unit tests, and they also revealed a sneaky bug which I fixed also in the same commit, please review carefully the changes and inform me if there are anymore changes to do

raman-m commented on src/Ocelot/DownstreamRouteFinder/UrlMatcher/UrlPathPlaceholderNameAndValueFinder.cs:
```diff
@@ -62,7 +57,7 @@ public Response<List<PlaceholderNameAndValue>> Find(string path, string query, s
                     else
                     {
-                        var placeholderValue = GetPlaceholderValue(pathTemplate, query, placeholderName, path, counterForPath + 1, '?');
+                        var placeholderValue = GetPlaceholderValue(pathTemplate, query, placeholderName, path, counterForPath, '?');
```
Wow! How did you figure out this `+1` issue? 😉 
And what was the impact?

AlyHKafoury replied:
When adding the query to unit tests I found that one of them was failing so tracing that I changed the plus 1 issue to handle specifically the no slash case only

raman-m commented on test/Ocelot.AcceptanceTests/RoutingTests.cs:
```diff
@@ -813,14 +813,14 @@ public void should_return_response_200_with_placeholder_for_final_url_path()
-        [InlineData("/downstream/{version}/test", "/upstream/{version}/test", "/upstream/v1/test", "/downstream/v1/test")]
-        [InlineData("/downstream/{version}/test", "/upstream/{version}/test", "/upstream/test", null)]
-        public void should_return_correct_downstream_when_omitting_ending_placeholder(string downstreamPathTemplate, string upstreamPathTemplate, string requestURL, string downstreamURL)
+        [InlineData("/downstream/test/{everything}", "/upstream/test/{everything}", "/upstream/test/1", "/downstream/test/1", "?p1=v1&p2=v2&something-else")]
+        [InlineData("/downstream/test/{everything}", "/upstream/test/{everything}", "/upstream/test/", "/downstream/test/", "?p1=v1&p2=v2&something-else")]
+        [InlineData("/downstream/test/{everything}", "/upstream/test/{everything}", "/upstream/test", "/downstream/test", "?p1=v1&p2=v2&something-else")]
+        [InlineData("/downstream/test/{everything}", "/upstream/test/{everything}", "/upstream/test123", null, null)]
+        [InlineData("/downstream/{version}/test/{everything}", "/upstream/{version}/test/{everything}", "/upstream/v1/test/123", "/downstream/v1/test/123", "?p1=v1&p2=v2&something-else")]
+        [InlineData("/downstream/{version}/test", "/upstream/{version}/test", "/upstream/v1/test", "/downstream/v1/test", "?p1=v1&p2=v2&something-else")]
+        [InlineData("/downstream/{version}/test", "/upstream/{version}/test", "/upstream/test", null, null)]
+        public void should_return_correct_downstream_when_omitting_ending_placeholder(string downstreamPathTemplate, string upstreamPathTemplate, string requestURL, string downstreamURL, string queryString)
```
Exactly it was my idea to fix... but it was too late for me yesterday's tonight 😸 
Thanks for fixing!

raman-m replied:
Seems our acceptance tests confirm that nothing broken in Catch-All feature, and we've introduced new empty placeholders for Catch-All too. Query String Catch-All should work too.
Great!

`^(?i)/contracts/\?.*$`

raman-m reviewed (commented):
Good job, Aly!
Thank you!

AlyHKafoury commented:
@raman-m I believe the branch is ready for merge my side, waiting for your confirmation or change requests

raman-m commented:
It's mostly ready, but not quite...

raman-m commented:
@AlyHKafoury  I've added [new these Routing tests](https://github.com/ThreeMammals/Ocelot/pull/1911/files#diff-99c1f697736fe2348d794d9911adf9dde5428205b123b99511fab187260ffd60R848-R868) and they failed.
Sorry, we have to fix them!... 😞 
This [theory](https://github.com/ThreeMammals/Ocelot/pull/1911/files#diff-99c1f697736fe2348d794d9911adf9dde5428205b123b99511fab187260ffd60R855) should check possible breaking changes for the [Catch All Query String](https://ocelot.readthedocs.io/en/latest/features/routing.html#catch-all-query-string) feature. Seems this feature is broken now, or tests should be fixed.

I'm going to debug today... It will be nice if you help me too!..

AlyHKafoury commented:
@raman-m Please leave the debugging to me thanks for adding the failings tests and add more failing tests if they are needed and I can debug  them to insure I understand all ocelot functionality

raman-m commented:
Thank you!
We've made aggressive refactoring in route matching logic...
So, Better to double check current [Routing](https://ocelot.readthedocs.io/en/latest/features/routing.html) sub-features.

AlyHKafoury commented:
@raman-m fixing the test cases now will submit commit today

AlyHKafoury commented:
@raman-m I added fixes for the unit tests, please review the added commits

raman-m reviewed (approved):
Ready for delivery! ✔️ 

Thanks, Aly!
Good job! 💪
